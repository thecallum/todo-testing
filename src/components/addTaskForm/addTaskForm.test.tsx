import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import AddTaskForm from "./addTaskForm";
import { expect } from '@jest/globals';

describe("AddTaskForm", () => {
  it("Renders component", () => {
    // Arrange
    const addTask = () => {};

    // Act
    const { asFragment } = render(<AddTaskForm addTask={addTask} />);

    // Assert
    expect(asFragment()).toMatchSnapshot();
  });

  it("When empty input, doesn't submit form", () => {
    // Arrange
    const addTask = jest.fn();

    const { container } = render(<AddTaskForm addTask={addTask} />);

    // Act
    const submitButton = container.querySelector("button.addTask-button") as Element;
    userEvent.click(submitButton);

    // Assert
    expect(addTask).toHaveBeenCalledTimes(0);
  });

  it("When valid input, does submit form", () => {
    // Arrange
    const addTask = jest.fn();
    const textInput = "foo";

    const { container } = render(<AddTaskForm addTask={addTask} />);

    // Act
    const input = container.querySelector("input.addTask-input") as Element;

    userEvent.click(input);
    userEvent.keyboard(textInput);

    const submitButton = container.querySelector("button.addTask-button") as Element;
    userEvent.click(submitButton);

    // Assert
    expect(addTask).toHaveBeenCalledTimes(1);
    expect(addTask).toHaveBeenCalledWith(textInput);
  });

  it("When form submitted, test input is cleared", () => {
    // Arrange
    const addTask = jest.fn();
    const textInput = "foo";

    const { container } = render(<AddTaskForm addTask={addTask} />);

    // Act
    const input = container.querySelector("input.addTask-input") as HTMLInputElement ;

    userEvent.click(input);
    userEvent.keyboard(textInput);

    const submitButton = container.querySelector("button.addTask-button") as Element;
    userEvent.click(submitButton);

    // Assert
    expect(addTask).toHaveBeenCalledTimes(1);

    expect(input.value).toEqual("");
  });
});

