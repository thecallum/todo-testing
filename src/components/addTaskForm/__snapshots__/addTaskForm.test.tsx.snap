// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTaskForm Renders component 1`] = `
<DocumentFragment>
  <form
    class="addTask-form"
  >
    <input
      class="addTask-input"
      data-cy="add-task-input"
      placeholder="Add a task"
      type="text"
      value=""
    />
    <button
      class="addTask-button"
      data-cy="add-task-button"
      type="submit"
    >
      Add
    </button>
  </form>
</DocumentFragment>
`;
